File: src/main/scala/MinimumWindowSubstring.java

public class MinimumWindowSubstring {
    public static String minWindow(String s, String t) {
         if(s == null || t == null || s.length() < t.length()) return "";
         int map[] = new int[128];
         for(int i = 0; i< t.length(); i++){
             map[t.charAt(i)] += 1;
         }
         int counter = t.length();
         int begin = 0,end = 0;
         int d = Integer.MAX_VALUE;
         int head = 0;
         while(end < s.length()){
             if(map[s.charAt(end)] > 0) counter -=1;
             map[s.charAt(end)] -= 1;
             end += 1;
             while(counter == 0){
                 if(end - begin < d) {
                     d = end - begin;
                     head = begin;
                 }
                 if(map[s.charAt(begin)] == 0) counter += 1;
                 map[s.charAt(begin)] += 1;
                 begin += 1;

             }
         }
         return d == Integer.MAX_VALUE ? "": s.substring(head, head + d);
    }

    public static void main(String ... args){
        System.out.println(minWindow("ADOBECODEBANC", "ABC"));
    }
}

File: src/main/scala/MinimumWindowSubstring.py
from collections import Counter
from collections import defaultdict


class Solution:
    def minWindow(self, s, t):
        if not s and not t:
            return t
        if not s or not t:
            return ""
        min_win = [float('inf'), None, None]
        chars = Counter(t)
        need = len(chars)
        win = defaultdict(int)
        left = 0
        formed = 0
        for right in range(len(s)):
            ch = s[right]
            win[ch] += 1
            if ch in chars and win[ch] == chars[ch]:
                formed += 1
            while left <= right and formed == need:
                if min_win[0] > right - left + 1:
                    min_win = [right - left + 1, left, right]
                win[s[left]] -= 1
                if s[left] in chars and win[s[left]] < chars[s[left]]:
                    formed -= 1
                left += 1
        formed, l, r = min_win
        return s[l: r + 1] if formed != float('inf') else ''


sol = Solution()
print(sol.minWindow("aa", "aa"))
print(sol.minWindow("bdab", "ab"))
print(sol.minWindow("a", "aa"))
print(sol.minWindow("ADOBECODEBANC", "ABC"))
